packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.java

```diff
@@ -0,0 +1,512 @@
+    }
+
+    if ("ALWAYS".equals(developerSetting)) {
+      return true;
+    } else if ("NEVER".equals(developerSetting)) {
+      return false;
+    } else if ("WIFI_ONLY".equals(developerSetting)) {
+      ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+      if (cm == null) {
+        Log.e(TAG, "Could not determine active network connection is metered; not checking for updates");
+        return false;
+      }
+      return !cm.isActiveNetworkMetered();
+    } else {
+      Log.e(TAG, "Invalid value " + developerSetting + " for expo.modules.updates.EXPO_UPDATES_CHECK_ON_LAUNCH in AndroidManifest; defaulting to ALWAYS");
+      return true;
+    }
+  }
+
+  private synchronized void finishTimeout() {
+    if (mTimeoutFinished) {
+      // already finished, do nothing
+      return;
+    }
+    mTimeoutFinished = true;
+    notify();
+  }
+
+  private void runReaper() {
+    AsyncTask.execute(() -> {
+      UpdatesDatabase database = getDatabase();
+      Reaper.reapUnusedUpdates(database, mUpdatesDirectory, getLaunchedUpdate(), mSelectionPolicy);
+      releaseDatabase();
+    });
+  }
+
+  public void relaunchReactApplication(Context context, Launcher.LauncherCallback callback) {
+    if (mReactNativeHost == null || mReactNativeHost.get() == null) {
+      callback.onFailure(new Exception("Could not reload application. Ensure you have passed the correct instance of ReactApplication into UpdatesController.initialize()."));
+      return;
+    }
+    final ReactNativeHost host = mReactNativeHost.get();
+
+    final String oldLaunchAssetFile = mLauncher.getLaunchAssetFile();
+
+    UpdatesDatabase database = getDatabase();
+    final LauncherWithSelectionPolicy newLauncher = new LauncherWithSelectionPolicy(mUpdatesDirectory, mSelectionPolicy);
+    newLauncher.launch(database, context, new Launcher.LauncherCallback() {
+      @Override
+      public void onFailure(Exception e) {
+        callback.onFailure(e);
+      }
+
+      @Override
+      public void onSuccess() {
+        mLauncher = newLauncher;
+        releaseDatabase();
+
+        final ReactInstanceManager instanceManager = host.getReactInstanceManager();
+
+        String newLaunchAssetFile = mLauncher.getLaunchAssetFile();
+        if (newLaunchAssetFile != null && !newLaunchAssetFile.equals(oldLaunchAssetFile)) {
+          // Unfortunately, even though RN exposes a way to reload an application,
+          // it assumes that the JS bundle will stay at the same location throughout
+          // the entire lifecycle of the app. Since we need to change the location of
+          // the bundle, we need to use reflection to set an otherwise inaccessible
+          // field of the ReactInstanceManager.
+          try {
+            JSBundleLoader newJSBundleLoader = JSBundleLoader.createFileLoader(newLaunchAssetFile);
+            Field jsBundleLoaderField = instanceManager.getClass().getDeclaredField("mBundleLoader");
```

Comment:
I am really afraid of this solution. This is something that should be thoughtfully tested on each react native update, and imposes huge risk of failure in the future.

I think we should either get rid of this functionality at all, or at least consider publishing it as separate module. This way we would remember it is risky and potentially discourage people from using it?

Reply:
Sending a PR upstream to expose this field seems like the simplest solution in the long run. We can drop support for old versions of RN as needed.

Reply:
FWIW -- this is what react-native-code-push does, as well. I think we have a good chance of upstreaming a change, though, and will scope that for next cycle.